js/src/logger.ts

```diff
@@ -1460,23 +1478,43 @@ export class Experiment {
   public kind: "experiment" = "experiment";
 
   constructor(
-    project: RegisteredProject,
-    id: string,
-    name: string,
+    lazyMetadata: Promise<ProjectExperimentMetadata>,
     dataset?: Dataset
   ) {
     this.finished = false;
 
-    this.project = project;
-    this.id = id;
-    this.name = name;
+    this.lazyMetadata = lazyMetadata;
     this.dataset = dataset;
     this.bgLogger = new BackgroundLogger();
     this.lastStartTime = getCurrentUnixTimestamp();
 
     unterminatedObjects.addUnterminated(this, iso.getCallerLocation());
   }
 
+  public get id(): Promise<string> {
```

Comment:
Technically this is a breaking change as we have changed these from bare values to promises. Not sure there's a good way around this unless we make the lazy init opt-in for all of these objects and do template parameterize everything to let people turn it on. But I feel like lazy init is the correct default so maybe we just eat the one-time cost?

Reply:
I think it's likely fine, but I'm not sure there's much value in lazily initializing these objects (specifically, experiments and datasets), either, since the `init` functions are already async?

Reply:
If we keep this, then when we land this change, we should just send a notification to customers in case they rely on this (although AFAIK, most, if not all, have moved to the `Eval` framework).

Reply:
Yeah I think I'm still leaning towards landing it. Seems like a nice feature, and I imagine the breakage is relatively benign. We do have https://linear.app/braintrustdata/issue/BRA-119/experiment-and-dataset-initialization-time-should-be-instant, which this solves (I can do a similar thing for python without changing the API at all).

Reply:
SGTM!